### PR TITLE
DATACOUCH-615 Use query in removeByQuery

### PR DESCRIPTION
**Pull request by haal**
Use provided query instead of hardcoded statement.

- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] There is a ticket in the bug tracker for the project in our [JIRA](https://jira.spring.io/browse/DATACOUCH).
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@haal Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-data-couchbase&pullRequestId=264)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-data-couchbase&pullRequestId=264) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@haal Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-data-couchbase&pullRequestId=264)!

File: src/main/java/org/springframework/data/couchbase/core/ReactiveRemoveByQueryOperationSupport.java [hunk 1]
```diff
@@ -75,7 +68,7 @@ public ReactiveRemoveByQueryOperationSupport(final ReactiveCouchbaseTemplate tem
 								return throwable;
 							}
 						}).flatMapMany(ReactiveQueryResult::rowsAsObject)
-						.map(row -> new RemoveResult(row.getString("id"), row.getLong("cas"), Optional.empty()));
+						.map(row -> new RemoveResult(row.getString("__id"), row.getLong("__cas"), Optional.empty()));
```
Comment (mikereiche) [inline src/main/java/org/springframework/data/couchbase/core/ReactiveRemoveByQueryOperationSupport.java:71]: The names (__id, __cas) should come from TemplateUtils, although they seem to be hard-coded in several other places.

File: src/main/java/org/springframework/data/couchbase/core/query/Query.java [hunk 2]
```diff
@@ -254,6 +254,16 @@ public String toN1qlString(ReactiveCouchbaseTemplate template, Class domainClass
 		return statement.toString();
 	}
 
+	public String toN1qlRemoveString(ReactiveCouchbaseTemplate template, Class domainClass) {
+		StringBasedN1qlQueryParser.N1qlSpelValues n1ql = getN1qlSpelValues(template, domainClass, false);
+		final StringBuilder statement = new StringBuilder();
+		appendString(statement, n1ql.delete); // delete ...
+		appendWhereString(statement, n1ql.filter); // typeKey = typeValue
+		appendWhere(statement, null); // criteria on this Query
```
Comment (mikereiche) [inline src/main/java/org/springframework/data/couchbase/core/query/Query.java:262]: If the second arg to appendWhere() is  new int[] { 0 }   (as in toN1qlSelectString), the variables in the criteria will be parameterized for efficiency.

**Review by mikereiche — APPROVED** [review]

File: src/main/java/org/springframework/data/couchbase/core/ReactiveRemoveByQueryOperationSupport.java (hunk 1 above)
Comment (haal) [reply src/main/java/org/springframework/data/couchbase/core/ReactiveRemoveByQueryOperationSupport.java]: Fixed

File: src/main/java/org/springframework/data/couchbase/core/query/Query.java (hunk 2 above)
Comment (haal) [reply src/main/java/org/springframework/data/couchbase/core/query/Query.java:262]: Initially I had the same idea but in that case positional parameters in query will not be replaced with actual values. Also, it seems in current code `findByQuery(clazz).matching(query)` method is not behaving as expected (added new test case in `CouchbaseTemplateQueryIntegrationTests` in this PR which is failing).

Comment (daschl) [reply src/main/java/org/springframework/data/couchbase/core/query/Query.java:262]: I don't think we should submit a pr with a failing test - maybe submit it in a different PR so we can work on a fix alongside it?

Comment (haal) [reply src/main/java/org/springframework/data/couchbase/core/query/Query.java:262]: Yes, I can create failing test in separate PR. The question is how to handle comment from @mikereiche where he proposes to use `new int[] {0}` which does not work at the moment?

Comment (haal) [reply src/main/java/org/springframework/data/couchbase/core/query/Query.java:262]: New bug created https://jira.spring.io/browse/DATACOUCH-619. 
Resolving initial comment without applying `new int[] {0}` as it is not working as expected.

**Review by daschl — APPROVED** [review]

**Pull request merged**